src/server/routes/api/document/index.js

```diff
@@ -1446,23 +1462,37 @@ http.get('/text/:id', function( req, res, next ){
     .catch( next );
 });
 
+const searchRelatedPapers = ( doc, interactionId ) => {
+  let article = doc.article();
+  if ( !article ) {
+    return [];
+  }
+
+  let templates;
+  if ( interactionId ) {
+    let intn = doc.get( interactionId );
+    templates = [ intn.toSearchTemplate() ];
+  }
+
+  templates = doc.toSearchTemplates();
+
+  let obj = { templates, article };
+  return indra.searchDocuments( obj );
```

Comment:
I was thinking that we could do that mapping while saving to database. Here (https://github.com/PathwayCommons/factoid/pull/727/files#diff-9d36d49f610f5307b10c657fc6e1c3a8R87) we have the ``docId`` property we can add ``elId`` as well which may not be set or just null when we work on the whole document.

Reply:
I'm just indicating that currently, there's no way to match a `template` with an interaction...

Reply:
@jvwong Do you mean that this function will never be called with ``interactionId`` parameter is set? 

If so, in the previous discussion we wanted to handle the query for a single interaction of a document as well in the related service end point. Actually, currently I just implemented the trigger for a whole document. However, I was assuming that the function will be called for an interaction as well though I do not know how/when it will be triggered. Will it not be called with ``interactionId`` parameter?

Reply:
Let's assume that the related papers will be fetched once, at the time the user submits. The `doc.toSearchTemplates` will aggregate results for all interactions. Imagine a user clicks on one interaction in the 'explorer' view: It would be nice to filter the papers for only that interaction. 

> Will it not be called with interactionId parameter?

I can't see the `interactionId` being used in the case above.  Maybe there's another case?

Reply:
@jvwong @maxkfranz 

Considering what Jeff just mentioned as well my understanding is like the following:

Actually there must be some different part of mechanism. Item 1 is a must and we must choose either one of 2 or 3. 4 may or may not be needed.

1) Query indra and other services in server side and store the results in database.
2) Create an end point for the client side explore view queries that will read from what we write to database after querying indra etc. The ``interactionId`` parameter that we are discussing may be needed  here instead of the existing service end point.
3) Do not create an end point for explore view queries. Instead of doing that store the results in ``interaction`` and ``document`` objects. Then use them from that objects directly in explore view.
4) The service end-point that we currently have (that actually triggers (1)). This service is not supposed to be called from explore view directly. But it may be triggered sometimes. The same job is done when a factoid is submitted without needing the end point. The end point may be needed if we want to have more triggers like triggering the related papers search from the command-line or from an npm command. Based on what I just learned this would not need the ``interactionId`` parameter.

Am I right on this points? Which one of 2, 3, 4 must exists?

Reply:
4 does not seem necessary. I don't have a preference for 2/3.

Reply:
The rest endpoints are useful for testing.  It's not useful for the ui, since it would always just read the model as per (3).  If you think that that routes would be useful to you for testing, then it makes sense to have them.

Reply:
I guess it may be useful in testing so I can keep it at least for now.

Reply:
@maxkfranz I added related papers data to Document and Element tables of database. What would be the proper way of attaching that data to the Interaction and Document objects? Should I make some changes in Syncer?

Reply:
Depending on the answer here I can revert the changes I made to store the relatedPapers info in Document and Element tables and I can store them in another RelatedPapers table. I did that only for the Documents in the past commits.

Reply:
Let's not create new database tables.  That introduces more complexity.  The related papers are always associated with particular documents or interactions, so that data should be stored under documents or interactions.  We're using a NoSQL database.  There's no need to create tables in a SQL style.

It's fine to just use something simple like `this.syncher.set('relatedPapers', myPapers)` and this.syncher.get('relatedPapers')` in the `Document`.  There are examples of other fields for `Document` and `Interaction` that use `this.syncher` that you can use as a reference.

Reply:
@maxkfranz Okay I did like that. However, that brought a question to my mind. I am updating the database for the related papers without touching the model (https://github.com/PathwayCommons/factoid/pull/727/files#diff-9d36d49f610f5307b10c657fc6e1c3a8R134). I guess I can do the same thing by using ``document.relatedPapers(data)`` and ``interaction.relatedPapers(data)''. Is there a need for using model/syncher instead of making direct db operations there for this case. I see that for tweeting the db is directly updated as well.

Reply:
Looks like the answer to this question is already available somewhere below.

Reply:
Yes, the twitter stuff should technically also be using the OO model.  However, it's a bit more important in the case of related papers, since the related papers feature uses the model semantics and the twitter stuff doesn't.